code/modules/ninja/suit/ninjaDrainAct.dm

```diff
@@ -107,37 +107,48 @@
 
 	return drain_total
 
-//RDCONSOLE//
-/obj/machinery/computer/rdconsole/ninjadrain_act(obj/item/clothing/suit/space/space_ninja/ninja_suit, mob/living/carbon/human/ninja, obj/item/clothing/gloves/space_ninja/ninja_gloves)
+//RD SERVER//
+/obj/machinery/rnd/server/master/ninjadrain_act(obj/item/clothing/suit/space/space_ninja/ninja_suit, mob/living/carbon/human/ninja, obj/item/clothing/gloves/space_ninja/ninja_gloves)
 	if(!ninja_suit || !ninja || !ninja_gloves)
 		return INVALID_DRAIN
 
 	. = DRAIN_RD_HACK_FAILED
 
-	to_chat(ninja, span_notice("Hacking \the [src]..."))
-	AI_notify_hack()
+	// If the traitor theft objective is still present, this will destroy it...
+	if(source_code_hdd)
+		to_chat(ninja, span_notice("Hacking \the [src]..."))
+		AI_notify_hack()
+		to_chat(ninja, span_notice("Encrypted source code detected. Overloading storage device..."))
+		if(do_after(ninja, 30 SECONDS, target = src) && src)
+			destroy_source_code_hdd()
+			to_chat(ninja, span_notice("Sabotage complete. Storage device overloaded."))
+			var/datum/antagonist/ninja/ninja_antag = ninja.mind.has_antag_datum(/datum/antagonist/ninja)
+			if(!ninja_antag)
+				return
+			var/datum/objective/research_secrets/objective = locate() in ninja_antag.objectives
+			if(objective)
+				objective.completed = TRUE
+		return
 
-	if(stored_research)
-		to_chat(ninja, span_notice("Copying files..."))
-		if(do_after(ninja, ninja_suit.s_delay, target = src) && src)
-			stored_research.copy_research_to(ninja_suit.stored_research)
-	to_chat(ninja, span_notice("Data analyzed. Process finished."))
+	// Otherwise, revert to the parent proc which drains research points instead.
+	return ..()
 
-//RD SERVER//
 /obj/machinery/rnd/server/ninjadrain_act(obj/item/clothing/suit/space/space_ninja/ninja_suit, mob/living/carbon/human/ninja, obj/item/clothing/gloves/space_ninja/ninja_gloves)
 	if(!ninja_suit || !ninja || !ninja_gloves)
 		return INVALID_DRAIN
 
 	. = DRAIN_RD_HACK_FAILED
 
-	to_chat(ninja, span_notice("Hacking \the [src]..."))
-	AI_notify_hack()
-
-	if(stored_research)
-		to_chat(ninja, span_notice("Copying files..."))
-		if(do_after(ninja, ninja_suit.s_delay, target = src) && src)
-			stored_research.copy_research_to(ninja_suit.stored_research)
-	to_chat(ninja, span_notice("Data analyzed. Process finished."))
+	to_chat(ninja, span_notice("Research notes detected. Corrupting data..."))
+	if(do_after(ninja, 30 SECONDS, target = src) && src)
```

Comment:
Early return, change to `!QDELETED(src)` (or nothing, since target is already checked in `do_after`)

Reply:
Done. Feel the early return makes code less readable though.